### PR TITLE
GH-95289: Always call uncancel() when parent cancellation is requested

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-95289

Co-Authored-By: Guido van Rossum <guido@python.org>

### CODE REVIEW

File: Lib/asyncio/taskgroups.py [hunk 1]
```diff
@@ -54,21 +54,22 @@ async def __aenter__(self):
...
-                if self._parent_cancel_requested and not self._parent_task.uncancel():
-                    # Do nothing, i.e. swallow the error.
-                    pass
-                else:
-                    propagate_cancellation_error = exc
+        propagate_cancellation_error = \
+            exc if et is exceptions.CancelledError else None
+        if self._parent_cancel_requested:
+            # If this flag is set we *must* call uncancel().
+            if self._parent_task.uncancel() == 0:
```
Comment (graingert) [inline Lib/asyncio/taskgroups.py:67]: there's now an issue where `uncancel() != 0` and there's no CancelledError to propagate:

```python
import asyncio
import contextlib

class CustomException(Exception):
    pass

async def raise_exc():
    raise CustomException

@contextlib.asynccontextmanager
async def database():
    try:
        yield
    finally:
        raise CustomException


async def main():
    try:
        async with asyncio.timeout(0.1):
            try:
                async with asyncio.TaskGroup():
                    async with database():
                        await asyncio.sleep(10)
            except* CustomException:
                 print("custom exception happened!")
    except TimeoutError as e:
        print("timed out correctly")
    else:
        assert False, "failed to timeout"

asyncio.run(main())
```

**Review by graingert — CHANGES_REQUESTED** [review]

File: Lib/asyncio/taskgroups.py (hunk 1 above)
Comment (kumaraditya303) [reply Lib/asyncio/taskgroups.py:67]: It is failing even on main without this PR so is unrelated, you should probably create a new issue for this.

Comment (graingert) [reply Lib/asyncio/taskgroups.py:67]: this one is happens on both main and this PR
I'd expect it to print "the customer lost data, we need to phone them up to fix it"


```python
import asyncio
import contextlib
import logging

logger = logging.getLogger(__name__)

class CustomerDataCorrupted(Exception):
    pass

async def raise_exc():
    raise CustomerDataCorrupted

@contextlib.asynccontextmanager
async def database():
    try:
        yield
    finally:
        print("customer error happened, but we don't have the context to deal with it here")
        raise CustomerDataCorrupted

async def with_database():
    async with database():
        await asyncio.sleep(10)


async def main():
    try:
        async with asyncio.timeout(0.1):
            try:
                async with asyncio.TaskGroup() as tg:
                    tg.create_task(with_database())
            except* CustomerDataCorrupted:
                 print("the customer lost data, we need to phone them up to fix it")
    except TimeoutError:
        pass

asyncio.run(main())
```

Comment (gvanrossum) [reply Lib/asyncio/taskgroups.py:67]: You are being a bit dramatic, Thomas. Both examples contain the same bug -- they have a context manager that accidentally replaces `CancelledError` with some other exception. The fix is something like
```
        if not isinstance(sys.exc_info()[1], asyncio.CancelledError):
            raise CustomException
```

File: Lib/test/test_asyncio/test_taskgroups.py [hunk 2]
```diff
@@ -741,6 +741,35 @@ async def coro2(g):
...
+
+        async def main():
+            task = asyncio.current_task()
+            try:
+                async with taskgroups.TaskGroup() as tg:
+                    async with database():
+                        tg.create_task(raise_exc())
+                        await asyncio.sleep(1)
+            except* CustomException:
+                self.assertEqual(task.cancelling(), 0)
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_taskgroups.py]: I find this test/example a bit confusing. It's clear that `CustomException` is raised but it's less clear *which* `CustomException` is being caught: the original one, raised by `raise_exc()`, or the one raised by the context manager. The answer is both -- and to verify that, we could add this:
```suggestion
            try:
                async with taskgroups.TaskGroup() as tg:
                    async with database():
                        tg.create_task(raise_exc())
                        await asyncio.sleep(1)
            except* CustomException as err:
                self.assertEqual(task.cancelling(), 0)
                self.assertEqual(len(err.exceptions), 2)

```
(One instance is the `raise_exc()` task, the other is the parent task being cancelled in the sleep.)

File: Lib/asyncio/taskgroups.py [hunk 3]
```diff
@@ -54,21 +54,22 @@ async def __aenter__(self):
...
+                # If there are no pending cancellations left,
+                # don't propagate CancelledError.
+                propagate_cancellation_error = None
```
Comment (gvanrossum) [inline Lib/asyncio/taskgroups.py:70]: If we want to fix Thomas's example, we can probably fix it like this:
```suggestion
                propagate_cancellation_error = None
            elif et is not exceptions.CancelledError:
                # If the exception is not CancelledError,
                # replace it with CancelledError.
                propagate_cancellation_error = exceptions.CancelledError()
```

Comment (gvanrossum) [reply Lib/asyncio/taskgroups.py:70]: Alas, not. It seems Thomas's example is not related to `uncancel()` -- that example doesn't even use `await` in the body of the `async with TaskGroup()` block, so `_parent_cancel_requested` is never set, and `uncancel()` *should* not be called.

That example needs to check whether one of the subtasks task was cancelled but not by us. There's no way we can fix that before RC1, but I also don't see that it is a fatal flaw -- it's a bug we can fix in 3.11.1 by adding more bookkeeping to `TaskGroup`.

**Comment by gvanrossum** [comment]
We're getting very close to the RC1 release date (this Friday!). I'd like to get this in. @graingert can you bless this as strictly better than what we had before?

File: Lib/asyncio/taskgroups.py (hunk 1 above)
Comment (kumaraditya303) [reply Lib/asyncio/taskgroups.py:67]: Marking as resolved as it is unrelated, also you should post unrelated issues which fail even on main to the issue not PRs.

File: Lib/asyncio/taskgroups.py (hunk 3 above)
Comment (kumaraditya303) [reply Lib/asyncio/taskgroups.py:70]: That can be dealt in another PR as "usual bug", it needs not to block this one.

**Comment by kumaraditya303** [comment]
> We're getting very close to the RC1 release date (this Friday!). I'd like to get this in


I agree, I would like to get this in ASAP.

File: Lib/asyncio/taskgroups.py (hunk 3 above)
Comment (graingert) [reply Lib/asyncio/taskgroups.py:70]: I've made a less dramatic version of what I think is this issue: 
https://gist.github.com/graingert/e8ddd68a0513bd7c02531a77a79cdacd

**Comment by graingert** [comment]
> We're getting very close to the RC1 release date (this Friday!). I'd like to get this in. @graingert can you bless this as strictly better than what we had before?

This now follows the five (six?) rules of uncancellation so I think therefore it's strictly better. However I can't say for sure. It looks to me like TaskGroup was relying on `task.cancelling() >= 1` traveling up the stack to communicate with parent scopes without anticipating that it also travels down the rest of the function body.

**Pull request merged**